compiler/resolution/functionResolution.cpp

```diff
@@ -11267,26 +11267,66 @@ static void checkSpeciallyNamedMethods() {
   static const std::unordered_map<const char*, InterfaceSymbol*> reservedNames = {
     { astr("hash"), gHashable },
     { astr("enterContext"), gContextManager },
-    { astr("exitContext"), gContextManager },
+    { astr("exitContext"),  gContextManager },
+    { astr("serialize"),    gWriteSerializable },
+    { astr("deserialize"),  gReadDeserializable },
+    { astr("init"),         gInitDeserializable },
   };
 
   SpecialMethodMap flagged;
 
   for_alive_in_Vec(FnSymbol, fn, gFnSymbols) {
     if (!fn->isMethod()) continue;
     if (fn->isCompilerGenerated() || fn->hasFlag(FLAG_FIELD_ACCESSOR)) continue;
-    // The parent class must have this function, and would've been flagged.
-    // the user will already get the warning.
-    if (fn->hasFlag(FLAG_OVERRIDE)) continue;
     auto reservedIter = reservedNames.find(fn->name);
     if (reservedIter == reservedNames.end()) continue;
+    auto found = reservedIter->second;
+
+    if (found == gHashable || found == gContextManager) {
+      // The parent class must have this function, and would've been flagged.
+      // the user will already get the warning.
+      if (fn->hasFlag(FLAG_OVERRIDE)) continue;
+    }
 
     auto receiverType = fn->getReceiverType();
     auto at = getBaseTypeForInterfaceWarnings(receiverType);
-    if (!at || (reservedIter->second == gHashable && at == dtObject)) continue;
+    if (!at || (found == gHashable && at == dtObject)) continue;
+
+    if (at->symbol->hasFlag(FLAG_ITERATOR_RECORD) &&
+        found == gWriteSerializable) continue;
 
-    auto key = SpeciallyNamedMethodKey(reservedIter->second, at);
+    int n = fn->numFormals();
+    if (fn->isInitializer() &&
+        !(n >= 4 &&
+        strcmp(fn->getFormal(n-1)->name, "reader") == 0 &&
+        strcmp(fn->getFormal(n)->name, "deserializer") == 0)) {
+      // Skip initializers that don't match the signature closely enough
+      continue;
+    }
+
+    if (found == gWriteSerializable) {
+      bool matches = n == 4 &&
+                     strcmp(fn->getFormal(3)->name, "writer") == 0 &&
+                     strcmp(fn->getFormal(4)->name, "serializer") == 0;
+      if (!matches) continue;
+    }
+
+    if (found == gReadDeserializable) {
+      bool matches = n == 4 &&
+                     strcmp(fn->getFormal(3)->name, "reader") == 0 &&
+                     strcmp(fn->getFormal(4)->name, "deserializer") == 0;
+      if (!matches) continue;
+    }
+
+    auto key = SpeciallyNamedMethodKey(found, at);
     flagged[key].speciallyNamedMethods[fn->name] = fn;
+
+    if (found == gWriteSerializable ||
+        found == gReadDeserializable ||
+        found == gInitDeserializable) {
```

Comment:
could there be something like `isSerdeAspectInterface` for this three-way disjunction?